go-fuzz-build/cover.go

```diff
@@ -849,9 +849,10 @@ func (f *File) newCounter(start, end token.Pos, numStmt int) ast.Stmt {
 		},
 		Index: idx,
 	}
-	return &ast.IncDecStmt{
-		X:   counter,
-		Tok: token.INC,
+	return &ast.AssignStmt{
+		Lhs: []ast.Expr{counter},
+		Tok: token.ASSIGN,
+		Rhs: []ast.Expr{ &ast.BasicLit{Kind: token.INT, Value: "1"} },
```

Comment:
Is this required? Or what's the motivation behind this change?

Reply:
Because otherwise each unique *amount* of branches executed registers as a new coverage signal. This leads to almost every new input getting added to the corpus.

Reply:
go-fuzz can operate in two modes (following AFL, as I understand it): either looking at which basic blocks are reached, or looking at a quantized count of how many times basic blocks are reached. See func `roundUpCover` and this quote from the [AFL whitepaper](http://lcamtuf.coredump.cx/afl/technical_details.txt): 

> In addition to detecting new tuples, the fuzzer also considers coarse tuple
hit counts. These are divided into several buckets:
>
>  1, 2, 3, 4-7, 8-15, 16-31, 32-127, 128+

But I'm still new here, so I might have gotten this wrong.

Reply:
In any case, if this is what libFuzzer needs, this behavior should probably be switched by flagLibFuzzer.

Reply:
I am still don't understand the full picture -- libfuzzer should support counters, but maybe it doesn't support them via __libfuzzer_extra_counters , or maybe the problem is that you dropped:
```
		for i := range CoverTab {	
			CoverTab[i] = 0	
		}
```
before each test.

Try to zero CoverTab before each test. If it does not help, then let's just switch between increments and setting to 1 depending on -libfuzzer flag.
But do we need to zero CoverTab anyway?... What does libfuzzer expect here?

Reply:
I'm afraid that using actual counters (used in standard go-fuzz, and as opposed to a binary CoverTab as implemented in my PR), will lead to a "coverage explosion".

For instance:

```go
for i := 0; i < input1; i++ {
        /* CoverTab updating inserted by go-fuzz here */
	for j := 0; j < input2; j++ {
                /* CoverTab updating inserted by go-fuzz here */
		for k := 0; k < input3; k++ {
                        /* CoverTab updating inserted by go-fuzz here */
			for l := 0; l < input4; l++ {
                            /* CoverTab updating inserted by go-fuzz here */
			}
		}
	}
}
```

With the standard go-fuzz CoverTab updating mechanism (incrementing the counter), this would lead to 2^32 different coverage signals, leading to 2^32 different files written to the corpus directory. This is obviously counter-productive since every unique input leads to new "coverage".

When we use my code that sets CoverTab to 1 if a branch is hit, and leaves it 0 as long it is not, the max code coverage is 4 instead of 2^32, which seems a lot more useful, especially if you consider complex targets that have enough branches of their own to give off a useful coverage signal.

I'm not a proponent of zeroing CoverTab each iteration because this incurs an unnecessary speed penalty that can be simply avoided by doing ```CoverTab[N] = 1``` in libFuzzer mode.

So I'll just make changes to the instrumentation code to insert my customized CoverTab updater.

In the future we can look at using a hybrid in the form of using 2 or 3 bits of coverage data per branch instead of 8 bits (standard go-fuzz) or 1 bit (my PR), depending on A/B testing.